src/Mod/Part/App/TopoShapeExpansion.cpp

```diff
@@ -2362,4 +2398,230 @@ bool TopoShape::fixSolidOrientation()
     return false;
 }
 
+TopoShape&
+TopoShape::makeElementBoolean(const char* maker, const TopoShape& shape, const char* op, double tolerance)
+{
+    return makeElementBoolean(maker, std::vector<TopoShape>(1, shape), op, tolerance);
+}
+
+
+TopoShape& TopoShape::makeElementBoolean(const char* maker,
+                                         const std::vector<TopoShape>& shapes,
+                                         const char* op,
+                                         double tolerance)
+{
+#if OCC_VERSION_HEX <= 0x060800
+    if (tolerance > 0.0) {
+        Standard_Failure::Raise("Fuzzy Booleans are not supported in this version of OCCT");
+    }
+#endif
+
+    if (!maker) {
+        FC_THROWM(Base::CADKernelError, "no maker");
+    }
+
+    if (!op) {
+        op = maker;
+    }
+
+    if (shapes.empty()) {
+        FC_THROWM(NullShapeException, "Null shape");
+    }
+
+    if (strcmp(maker, Part::OpCodes::Compound) == 0) {
+        return makeElementCompound(shapes, op, SingleShapeCompoundCreationPolicy::returnShape);
+    }
+    else if (boost::starts_with(maker, Part::OpCodes::Face)) {
+        std::string prefix(Part::OpCodes::Face);
+        prefix += '.';
+        const char* face_maker = 0;
+        if (boost::starts_with(maker, prefix)) {
+            face_maker = maker + prefix.size();
+        }
+        return makeElementFace(shapes, op, face_maker);
+    }
+    else if (strcmp(maker, Part::OpCodes::Wire) == 0) {
+        return makeElementWires(shapes, op);
+    }
+    else if (strcmp(maker, Part::OpCodes::Compsolid) == 0) {
+        BRep_Builder builder;
+        TopoDS_CompSolid Comp;
+        builder.MakeCompSolid(Comp);
+        for (auto& s : shapes) {
+            if (!s.isNull()) {
+                builder.Add(Comp, s.getShape());
+            }
+        }
+        setShape(Comp);
+        mapSubElement(shapes, op);
+        return *this;
+    }
+
+    if (strcmp(maker, Part::OpCodes::Pipe) == 0) {
+        if (shapes.size() != 2) {
+            FC_THROWM(Base::CADKernelError, "Not enough input shapes");
```

Comment:
This error seems off -- what if there are three?

Reply:
Hmm.  original source is assuming only ever two shapes for a pipe.  Which is consistent with the declaration of BRepOffsetAPI_MakePipe and the call on 2464.  Seems correct for this code, which can't work with other than 2.

Reply:
Then it's just the error message itself that's potentially misleading: if I selected three shapes, the error shouldn't say "Not enough input shapes" 😄.